### ISSUE TITLE
memory leaks on qemu-linux-user (docker/qemu-arm)

### ISSUE THREAD

**Issue by chen-charles** [chen-charles/psp2cldr#2]
Platform: Windows 11 `22000.469`  
When attempted to run `psp2cldr` on Docker / qemu-arm, a noticeable memory leak accumulates very quickly when threads are created.  

Narrowed down the issue further to the following sample program,  
```c
// compile with `gcc a.c -pthread`

#include <assert.h>
#include <pthread.h>

#define MAGIC_RETURN ((void *)42)

void *thread_main(void *arg)
{
    return MAGIC_RETURN;
}

int main(int argc, char *argv[])
{
    size_t i;
    for (i = 0;; i++)
    {
        pthread_t thread;
        assert(pthread_create(&thread, NULL, thread_main, NULL) == 0);
        void *ret;
        assert(pthread_join(thread, &ret) == 0);
        assert(ret == MAGIC_RETURN);
    }

    return 0;
}
```
A `pmap -XX` is attached ([a.log](https://github.com/chen-charles/psp2cldr/files/8020645/a.log)) for the above sample on `arm32v7/fedora:35`.  
Notice that the total mapped memory in vaddr space is `16066068 KiB`.  

Reproducible with multiple distributions from docker hub, including `arm32v7/alpine:3.15.0` (musl), `arm32v7/ubuntu:focal-20220113` (glibc), and `arm32v7/fedora:35` (glibc).  
Also reproducible on `qemu-arm version 6.1.0 (qemu-6.1.0-10.fc35)` with `gcc-arm-10.2-2020.11-x86_64-arm-none-linux-gnueabihf` [10.2-2020.11](https://developer.arm.com/-/media/Files/downloads/gnu-a/10.2-2020.11/binrel/gcc-arm-10.2-2020.11-x86_64-arm-none-linux-gnueabihf.tar.xz?revision=d0b90559-3960-4e4b-9297-7ddbc3e52783&la=en&hash=985078B758BC782BC338DB947347107FBCF8EF6B) from ARM's [GNU Toolchain for the A-profile Architecture](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-a/downloads).

**Comment by chen-charles** [comment]
Reproduced on `fedora 35` host: `:core-4.1-amd64:core-4.1-noarch`  
`qemu-arm version 6.1.0 (qemu-6.1.0-10.fc35)`  
`gcc-arm-10.2-2020.11-x86_64-arm-none-linux-gnueabihf` [10.2-2020.11](https://developer.arm.com/-/media/Files/downloads/gnu-a/10.2-2020.11/binrel/gcc-arm-10.2-2020.11-x86_64-arm-none-linux-gnueabihf.tar.xz?revision=d0b90559-3960-4e4b-9297-7ddbc3e52783&la=en&hash=985078B758BC782BC338DB947347107FBCF8EF6B)

**Comment by chen-charles** [comment]
Confimed this is a `qemu-linux-user` memory leak thanks to `valgrind --xtree-memory=full`.  
There are actually multiple sources of memory leaks, I have addessed them locally but it is probably worth a conversation in `qemu`'s mailing lists.

**Comment by chen-charles** [comment]
[qemu-project/qemu#866](https://gitlab.com/qemu-project/qemu/-/issues/866)

**Comment by chen-charles** [comment]
Unsophisticated patch to fix the memory leaks based off master@0a301624c2f4ced3331ffd5bce85b4274fe132af  
This patch only intends to address the issue in `qemu-arm`, will have to do per-platform investigation for other `linux-user` platforms.  

```diff
From 34cb3200bf99a841447f06cfdfe30ec7a9be41f4 Mon Sep 17 00:00:00 2001
From: Jianye Chen <jianye.chen@uwaterloo.ca>
Date: Thu, 10 Feb 2022 00:24:53 -0500
Subject: [PATCH] fixing qemu-arm memory leaks

leak 1: device_set_realized attaches (container_get(qdev_get_machine(), "/unattached") when we call cpu_copy at TARGET_NR_fork.
We effectively would now have an extra reference to the newly created CPU. Upon TARGET_NR_exit, we unref the thread CPU in hope that it will be collected, but because of this extra reference, it would be kept forever.

leak 2: ARMCPU->cpreg_[indexes|values|vmstate_indexes|vmstate_values]
init_cpreg_list allocates four pointers on ARMCPU, but no one is cleaning them up

leak 3: arm_cpu_register_gdb_regs_for_features
gdb_stub generates a large XML string and a uint32_t for each register
it also has a GDBRegisterState for each coprocessor that never get released

Signed-off-by: Jianye Chen <jianye.chen@uwaterloo.ca>
---
 gdbstub.c              | 11 +++++++++++
 include/exec/gdbstub.h |  1 +
 linux-user/syscall.c   |  1 +
 target/arm/cpu-qom.h   |  2 ++
 target/arm/cpu.c       | 13 +++++++++++++
 target/arm/gdbstub.c   | 12 ++++++++++++
 target/arm/helper.c    |  8 ++++++++
 target/arm/internals.h |  1 +
 8 files changed, 49 insertions(+)

diff --git a/gdbstub.c b/gdbstub.c
index 3c14c6a038..4b22695516 100644
--- a/gdbstub.c
+++ b/gdbstub.c
@@ -1011,6 +1011,17 @@ void gdb_register_coprocessor(CPUState *cpu,
     }
 }
 
+void gdb_cleanup_coprocessors(CPUState *cpu)
+{
+    GDBRegisterState *s = cpu->gdb_regs, *p;
+    while (s) {
+        p = s->next;
+        g_free(s);
+        s = p;
+    }
+    cpu->gdb_regs = NULL;
+}
+
 #ifndef CONFIG_USER_ONLY
 /* Translate GDB watchpoint type to a flags value for cpu_watchpoint_* */
 static inline int xlat_gdb_type(CPUState *cpu, int gdbtype)
diff --git a/include/exec/gdbstub.h b/include/exec/gdbstub.h
index a024a0350d..d939b0171d 100644
--- a/include/exec/gdbstub.h
+++ b/include/exec/gdbstub.h
@@ -83,6 +83,7 @@ typedef int (*gdb_set_reg_cb)(CPUArchState *env, uint8_t *buf, int reg);
 void gdb_register_coprocessor(CPUState *cpu,
                               gdb_get_reg_cb get_reg, gdb_set_reg_cb set_reg,
                               int num_regs, const char *xml, int g_pos);
+void gdb_cleanup_coprocessors(CPUState *cpu);
 
 /*
  * The GDB remote protocol transfers values in target byte order. As
diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index b9b18a7eaf..21fdf1bde1 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -8585,6 +8585,7 @@ static abi_long do_syscall1(void *cpu_env, int num, abi_long arg1,
             TaskState *ts = cpu->opaque;
 
             object_property_set_bool(OBJECT(cpu), "realized", false, NULL);
+            object_unparent(OBJECT(cpu));
             object_unref(OBJECT(cpu));
             /*
              * At this point the CPU should be unrealized and removed
diff --git a/target/arm/cpu-qom.h b/target/arm/cpu-qom.h
index a22bd506d0..45a05caa68 100644
--- a/target/arm/cpu-qom.h
+++ b/target/arm/cpu-qom.h
@@ -55,6 +55,7 @@ struct ARMCPUClass {
 
     const ARMCPUInfo *info;
     DeviceRealize parent_realize;
+    DeviceUnrealize parent_unrealize;
     DeviceReset parent_reset;
 };
 
@@ -72,6 +73,7 @@ struct AArch64CPUClass {
 
 void register_cp_regs_for_features(ARMCPU *cpu);
 void init_cpreg_list(ARMCPU *cpu);
+void cleanup_cpreg_list(ARMCPU *cpu);
 
 /* Callback functions for the generic timer's timers. */
 void arm_gt_ptimer_cb(void *opaque);
diff --git a/target/arm/cpu.c b/target/arm/cpu.c
index 5a9c02a256..f50c25a1db 100644
--- a/target/arm/cpu.c
+++ b/target/arm/cpu.c
@@ -1963,6 +1963,17 @@ static void arm_cpu_realizefn(DeviceState *dev, Error **errp)
     acc->parent_realize(dev, errp);
 }
 
+static void arm_cpu_unrealizefn(DeviceState *dev)
+{
+    ARMCPU *cpu = ARM_CPU(dev);
+    ARMCPUClass *acc = ARM_CPU_GET_CLASS(dev);
+
+    cleanup_cpreg_list(cpu);
+    arm_cpu_cleanup_gdb(cpu);
+
+    acc->parent_unrealize(dev);
+}
+
 static ObjectClass *arm_cpu_class_by_name(const char *cpu_model)
 {
     ObjectClass *oc;
@@ -2054,6 +2065,8 @@ static void arm_cpu_class_init(ObjectClass *oc, void *data)
 
     device_class_set_parent_realize(dc, arm_cpu_realizefn,
                                     &acc->parent_realize);
+    device_class_set_parent_unrealize(dc, arm_cpu_unrealizefn,
+                                      &acc->parent_unrealize);
 
     device_class_set_props(dc, arm_cpu_properties);
     device_class_set_parent_reset(dc, arm_cpu_reset, &acc->parent_reset);
diff --git a/target/arm/gdbstub.c b/target/arm/gdbstub.c
index ca1de47511..a81828dae8 100644
--- a/target/arm/gdbstub.c
+++ b/target/arm/gdbstub.c
@@ -503,3 +503,15 @@ void arm_cpu_register_gdb_regs_for_features(ARMCPU *cpu)
                              "system-registers.xml", 0);
 
 }
+
+void arm_cpu_cleanup_gdb(ARMCPU *cpu)
+{
+    CPUState *cs = CPU(cpu);
+
+    gdb_cleanup_coprocessors(cs);
+    
+    g_free(cpu->dyn_sysreg_xml.data.cpregs.keys);
+    g_free(cpu->dyn_sysreg_xml.desc);
+    g_free(cpu->dyn_svereg_xml.data.cpregs.keys);
+    g_free(cpu->dyn_svereg_xml.desc);
+}
diff --git a/target/arm/helper.c b/target/arm/helper.c
index b5f80988c9..ffec319ec8 100644
--- a/target/arm/helper.c
+++ b/target/arm/helper.c
@@ -283,6 +283,14 @@ void init_cpreg_list(ARMCPU *cpu)
     g_list_free(keys);
 }
 
+void cleanup_cpreg_list(ARMCPU *cpu)
+{
+    g_free(cpu->cpreg_indexes);
+    g_free(cpu->cpreg_values);
+    g_free(cpu->cpreg_vmstate_indexes);
+    g_free(cpu->cpreg_vmstate_values);
+}
+
 /*
  * Some registers are not accessible from AArch32 EL3 if SCR.NS == 0.
  */
diff --git a/target/arm/internals.h b/target/arm/internals.h
index 3f05748ea4..404f54e652 100644
--- a/target/arm/internals.h
+++ b/target/arm/internals.h
@@ -171,6 +171,7 @@ static inline int r14_bank_number(int mode)
 }
 
 void arm_cpu_register_gdb_regs_for_features(ARMCPU *cpu);
+void arm_cpu_cleanup_gdb(ARMCPU *cpu);
 void arm_translate_init(void);
 
 #ifdef CONFIG_TCG
-- 
2.25.1

```

**Comment by chen-charles** [comment]
[linux-user: un-parent OBJECT(cpu) when closing thread](https://gitlab.com/qemu-project/qemu/-/commit/6490d9aa62ef3cbbac2bf584fb0f3e737ab05e44)
https://lore.kernel.org/qemu-devel/20230106194451.1213153-1-richard.henderson@linaro.org/

fixed in `QEMU`; first contained in release `8.0.0`
